lib/src/isolate_manager/isolate_manager_io.dart

```diff
@@ -0,0 +1,128 @@
+  final _receivePort = ReceivePort();
+  final _errorPort = ReceivePort();
+  final _exitPort = ReceivePort();
+  // `_managerReadyCompleter` is a `Completer` that completes when the isolate
+  // is ready to receive tasks.
+  final Completer<void> _managerReadyCompleter = Completer<void>();
+  // `_isolateStatus` keeps track of the current status of the isolate.
+  IsolateStatus _isolateStatus = IsolateStatus.created;
+
+  // `start()` initializes the isolate and sets up listeners for errors and
+  // exit events. It waits for the `sendPort` from the isolate to be received
+  // before setting `_sendPort` and completing `_managerReadyCompleter`.
+  // If `start()` is called multiple times, an assertion error is thrown.
+  @override
+  Future<void> start() async {
+    assert(_isolateStatus == IsolateStatus.created, 'Isolate already started');
+    _isolateStatus = IsolateStatus.initializing;
+    final stream = _receivePort.asBroadcastStream();
+    resultStream = stream.where((message) => message is Result).cast<Result>();
+    _isolate = await Isolate.spawn(
+      _isolateFunction,
+      _receivePort.sendPort,
+      errorsAreFatal: false,
+      onError: _errorPort.sendPort,
+      onExit: _exitPort.sendPort,
+    );
+    _errorSubscription = _errorPort.listen((message) {
+      log('isolate manager error: $message');
+    });
+    _exitSubscription = _exitPort.listen((message) {
+      log('isolate manager exit: $message');
+      _errorSubscription.cancel();
+    });
+
+    final sendPortMessage = await stream.first;
+    _handleMessage(sendPortMessage);
+  }
+
+  /// Stops the isolate by killing it and cancelling error and exit event
+  /// listeners. Throws an assertion error if the isolate is not started.
+  @override
+  void stop() {
+    assert(_isolateStatus == IsolateStatus.initialized, 'Isolate not started');
+    _isolate.kill(priority: Isolate.immediate);
+    _errorSubscription.cancel();
+    _exitSubscription.cancel();
+    _isolateStatus = IsolateStatus.stopped;
+  }
+
+  /// Sends a task to the isolate, waits for the result, and returns the result.
+  /// Waits for _managerReadyCompleter to complete before sending the task.
+  /// Throws an error if the result contains an error.
+  @override
+  Future<FutureOr<T>> send<T>(
+    Response<dynamic> response,
+    FutureOr<T> Function(Response<dynamic>) mapper,
+  ) async {
+    await _managerReadyCompleter.future;
+    final task = Task(response.hashCode.toString(), response, mapper);
+
+    _sendPort.send(task);
+
+    final result =
+        await resultStream.where((event) => event.id == task.id).first;
+
+    if (result.error != null) {
+      throw result.error!;
+    }
+
+    return result.result;
+  }
+
+  /// Handles incoming messages from the isolate.
+  /// If the message is a [SendPort], sets the [_sendPort], completes the
+  /// [_managerReadyCompleter], and updates the [_isolateStatus].
+  void _handleMessage(dynamic message) {
+    if (message is SendPort) {
```

Comment:
let's switch to assert. We follow the contract that the first message should always be a sendport instance. Otherwise it is an expection state.

Reply:
Added it.